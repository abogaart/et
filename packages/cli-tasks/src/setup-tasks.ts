const listrInquirer = require('listr-inquirer');
const path = require('path');

export const Tasks = {
  choice(message: string, choices: string[], next: (path: string) => Promise<void>) {
    return listrInquirer([
      {
        type: 'list',
        name: 'choice',
        message,
        choices,
      },
    ], async (answers: any) => next(answers.choice));
  },

  confirm(message: string) {
    return {
      type: 'confirm',
      name: 'continue',
      message,
    };
  },

  async customPath(
    confirmMsg: string,
    pathLabel: string,
    next: (path: string) => Promise<void>,
    error: () => Promise<void>) {
    return listrInquirer([
      Tasks.confirm(confirmMsg),
      {
        when: (answers: any) => answers.continue,
        type: 'input',
        name: 'customPath',
        message: pathLabel,
        validate: (str: string) => {
          if (str && str.trim() !== '') {
            return path.isAbsolute(str) || 'Path should be absolute';
          }
          return 'Value is required';
        },
      },
    ], async (answers: any) => (answers.continue === false
      ? error()
      : next(answers.customPath.trim())));
  },
};
