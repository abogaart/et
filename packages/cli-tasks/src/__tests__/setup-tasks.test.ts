import { Tasks } from '../setup-tasks';

import listrInquirer = require('listr-inquirer');

jest.mock('listr-inquirer');

describe('setup tasks', () => {
  beforeEach(() => {
    listrInquirer.mockReset();
  });

  describe('choice', () => {
    const next = jest.fn();

    beforeEach(() => {
      next.mockReset();
      Tasks.choice('msg', ['choice1', 'choice2'], next);
    });

    it('invokes listr-inquirer with a choice list', () => {
      const [config] = listrInquirer.mock.calls[0];
      expect(config).toMatchSnapshot();
    });

    it('invokes next with the chosen value', async () => {
      const [, done] = listrInquirer.mock.calls[0];
      await done({ choice: 'return-value' });

      expect(next).toHaveBeenCalledWith('return-value');
    });
  });

  it('confirm', () => {
    expect(Tasks.confirm('confirm-msg')).toMatchSnapshot();
  });

  describe('customPath', () => {
    let config: any[];
    let error: () => Promise<void>;
    let next: (path: string) => Promise<void>;
    let done: (answers: any) => Promise<void>;

    beforeEach(async () => {
      error = jest.fn();
      next = jest.fn();

      await Tasks.customPath('confirm-msg', 'path-label', next, error);

      [[config, done]] = listrInquirer.mock.calls;
    });

    it('starts with a confirmation', async () => {
      const [confirm] = config;
      expect(confirm).toMatchSnapshot();
    });

    it('only shows the custom path input option if first step is confirmed', () => {
      const [, customPath] = config;
      expect(customPath.when({ continue: false })).toBe(false);
      expect(customPath.when({ continue: true })).toBe(true);
    });

    it('shows custom path input option', () => {
      const [, customPath] = config;
      expect(customPath).toMatchSnapshot();
    });

    it('validates that the input is not empty and an absolute path', () => {
      const [, customPath] = config;
      expect(customPath.validate('')).toBe('Value is required');
      expect(customPath.validate(' \n')).toBe('Value is required');
      expect(customPath.validate('path')).toBe('Path should be absolute');
      expect(customPath.validate('/path')).toBe(true);
    });

    it('invokes next with trimmed path when done', async () => {
      await done({ continue: true, customPath: '  /path\n' });
      expect(next).toHaveBeenCalledWith('/path');
    });

    it('invokes error if custom path is not set', async () => {
      await done({ continue: false });
      expect(error).toHaveBeenCalled();
    });
  });
});
