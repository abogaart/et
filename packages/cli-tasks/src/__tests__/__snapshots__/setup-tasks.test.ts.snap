// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setup tasks choice invokes listr-inquirer with a choice list 1`] = `
Array [
  Object {
    "choices": Array [
      "choice1",
      "choice2",
    ],
    "message": "msg",
    "name": "choice",
    "type": "list",
  },
]
`;

exports[`setup tasks confirm 1`] = `
Object {
  "message": "confirm-msg",
  "name": "continue",
  "type": "confirm",
}
`;

exports[`setup tasks customPath shows custom path input option 1`] = `
Object {
  "message": "path-label",
  "name": "customPath",
  "type": "input",
  "validate": [Function],
  "when": [Function],
}
`;

exports[`setup tasks customPath starts with a confirmation 1`] = `
Object {
  "message": "confirm-msg",
  "name": "continue",
  "type": "confirm",
}
`;
